test/integration/client/exec_test.go

```diff
@@ -0,0 +1,446 @@
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = []clientcmdapi.ExecEnvVar{
+					{
+						Name: outputEnvVar,
+						Value: fmt.Sprintf(`{
+							"kind": "ExecCredential",
+							"apiVersion": "client.authentication.k8s.io/v1beta1",
+							"status": {
+								"token": "%s"
+							}
+						}`, clientAuthorizedToken),
+					},
+				}
+				c.Username = "unauthorized"
+				c.Password = "unauthorized"
+			},
+			wantAuthorizationHeaderValues: []string{"Basic " + basicAuthHeaderValue("unauthorized", "unauthorized")},
+			wantCertificate:               &tls.Certificate{},
+			wantClientErrorPrefix:         "Unauthorized",
+		},
+		{
+			name: "good token with static auth bearer token favors static auth bearer token",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = []clientcmdapi.ExecEnvVar{
+					{
+						Name: outputEnvVar,
+						Value: fmt.Sprintf(`{
+							"kind": "ExecCredential",
+							"apiVersion": "client.authentication.k8s.io/v1beta1",
+							"status": {
+								"token": "%s"
+							}
+						}`, clientAuthorizedToken),
+					},
+				}
+				c.BearerToken = "some-unauthorized-token"
+			},
+			wantAuthorizationHeaderValues: []string{"Bearer some-unauthorized-token"},
+			wantClientErrorPrefix:         "Unauthorized",
+		},
+		{
+			// This is not the behavior we would expect, see
+			//   https://github.com/kubernetes/kubernetes/issues/99603
+			name: "good token with static auth cert and key favors exec plugin",
+			clientConfigFunc: func(c *rest.Config) {
+				c.ExecProvider.Env = []clientcmdapi.ExecEnvVar{
+					{
+						Name: outputEnvVar,
+						Value: fmt.Sprintf(`{
+							"kind": "ExecCredential",
+							"apiVersion": "client.authentication.k8s.io/v1beta1",
+							"status": {
+								"token": "%s"
+							}
+						}`, clientAuthorizedToken),
+					},
+				}
+				c.CertData = unauthorizedCert
+				c.KeyData = unauthorizedKey
+			},
+			wantAuthorizationHeaderValues: []string{"Bearer " + clientAuthorizedToken},
+			wantCertificate:               x509KeyPair([]byte(unauthorizedCert), []byte(unauthorizedKey), false),
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tmpDir, err := ioutil.TempDir("", "kubernetes-client-exec-test-plugin-dir-*")
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer os.RemoveAll(tmpDir)
+
+			var authorizationHeaderValues []string
+			clientConfig := rest.AnonymousClientConfig(result.ClientConfig)
+			clientConfig.ExecProvider = &clientcmdapi.ExecConfig{
+				Command: "testdata/exec-plugin.sh",
+				// TODO(ankeesler): move to v1 once exec plugins go GA.
+				APIVersion: "client.authentication.k8s.io/v1beta1",
+			}
+			clientConfig.Wrap(transport.WrapperFunc(func(rt http.RoundTripper) http.RoundTripper {
+				return roundTripperFunc(func(req *http.Request) (*http.Response, error) {
+					authorizationHeaderValues = req.Header.Values("Authorization")
```

Comment:
This could be called multiple times right?  Possibly concurrently?

Reply:
Makes sense. Committed at c5e14dc8fad1707fd6735bb6698447e37191a7ab, and then squashed everything together at 8834aa133c0007b5a6accff118511d2d88aaf344.